### DISCUSSION TITLE
R  in frontmatter with custom listing template

### DISCUSSION THREAD

**Discussion by nurfatimaj** [quarto-dev/quarto-cli#2202] [Q&A] [answered]
Hi!

I would like to create a listing item where some elements of the frontmatter depend on the output of R command like in the example below

```yaml
---
title: "title"
authors: 
  - name: "`r rmarkdown::yaml_front_matter(list.files(params[['project-path']], 'index.Rmd', full.names = TRUE))$author`"

params:
    project-path: '~/project/'
---
```

That is, I have a separate `project` folder, where I have all my work and output relevant to that project. I would like quarto to access this information when rendering the website, to minimize duplication. The idea is to access this way abstract and all relevant pdf outputs without having to copy paste them manually. 

I tried the above example and it works well when I am rendering that particular page. 
![image](https://user-images.githubusercontent.com/24235395/187322382-7d7af71f-5ea9-49fd-a29f-f5c8edeceea6.png)

But on the listing page, I get this

![image](https://user-images.githubusercontent.com/24235395/187322467-edebb473-4e4b-4cca-b021-f3e21f468143.png)

I am not sure, if there is a deeper issue why listings can't handle R code or is it an issue with my listing template. If latter, could you please let me know, how can I enable evaluation of R code in frontmatter yaml. Here is the portion of my template that prints the authors:

```ejs
<div class="quarto-title-authors">
    <ul class="listing-author">
        <% for (const athr of item.authors) {%>
            <% if (athr.url) { %>
                <li><a href="<%- athr.url %>"><%- athr.name %></a></li>
            <% } else { %>
                <li><%- athr.name %></li>
            <% } %>
        <% } %>
    </ul>
</div>
```

Thank you!

**Comment by mcanouil** [comment]
To use R code in YAML, you have to use `!expr` prefix (see <https://quarto.org/docs/computations/r.html#chunk-options>).

But the issue might also be related to <https://github.com/quarto-dev/quarto-cli/issues/1505>.

**Reply by nurfatimaj** [reply]
So, I noticed that about chunk options, but it doesn't seem to work for frontmatter yaml. Unless I am implementing it in a wrong way.

```yaml
title: "!expr rmarkdown::yaml_front_matter(list.files(params[['project-path']], 'index.Rmd', full.names = TRUE))$title"
params:
  project-path: '~/project/'
```
produces this
![image](https://user-images.githubusercontent.com/24235395/187337380-7a83bf12-17c4-45b7-9a4e-fdafc5d5c4ac.png)
both on the page of the given quarto document and on the listing page. 

I've also seen this discussion https://github.com/quarto-dev/quarto-cli/discussions/606, but was wondering if there are any new developments that allow me to do what I want.

**Reply by mcanouil** [reply]
No quotes around `!expr`.
Also, I don't fully understand your use case right know, what you seem to want feels highly strange.

Why use a so cumbersome solution to read metadata from an Rmarkdown (not a Quarto document?)? Why not use <https://quarto.org/docs/authoring/variables.html#var>?

**Reply by nurfatimaj** [reply]
I did try using `!expr` in frontmatter yaml without double quotes and that just doesn't render at all. 

Let me try to explain again what I would like to achieve. Let's say I have two folders on my computer called `paper` and `website`. 

```
| paper
|--index.Rmd
|--...
| website
|-- index.qmd
|-- listing.qmd
|-- research
|----paperA
|------ index.qmd
```

The `paper` folder lives outside of `website` folder. All the work is done inside the `paper` folder, like data analysis, producing drafts and slides. On the website I would like to report some information about the papers: title, authors, abstract. All of this information is contained in `paper/index.Rmd`. Currently, I am copying all of these by hand. What I would like to achieve is that I provide a path to `paper` folder  in `website/research/paperA/index.qmd` frontmatter and it would collect the required information from `paper/index.Rmd` automatically. 

If I understand correctly, there can only be one file called `_variables.yml` in the website. Therefore, I think providing project path inside the specific quarto document is more transparent, seeing that hopefully I will have more than a couple of papers in my life. Nevertheless, even inside `_variables.yml` I would still like to fetch the information from `paper/index.Rmd`, which does run into the same issue. 

However strange it might seem, would this kind of stuff be possible? Well, I know that it is possible if I simply want to render `website/research/paperA/index.qmd`. But would it be possible to have the same result when I try to render `website/listing.qmd`, which collects everything inside `website/research` folder?

P.S. Just to anticipate a possible answer. No, I cannot move `paper` folder inside `website` folder. Wouldn't have had a headache otherwise 😊

**Reply by cscheid** [reply]
You need double quotes _after_ `!expr`

This is incorrect syntax:

```
title: "!expr rmarkdown::yaml_front_matter(list.files(params[['project-path']], 'index.Rmd', full.names = TRUE))$title"
params:
  project-path: '~/project/'
```

This is the appropriate syntax:

```
title: !expr "rmarkdown::yaml_front_matter(list.files(params[['project-path']], 'index.Rmd', full.names = TRUE))$title"
params:
  project-path: '~/project/'
```

YAML is not intuitive, and more advanced uses like this will require unfortunately that you're comfortable with these sharper corners.

**Reply by nurfatimaj** [reply]
Thank you for a reply! I tried your suggestions and still doesn't work. I thought perhaps it's something wrong with my R function so I tried rendering the following dummy page
```
---
title: !expr "1+1"
editor: visual
---
```
And I keep getting this error: "ERROR: TypeError: title2.trim is not a function".

I tried also using `as.character(1+1)` in this dummy page, removing the quotes, quoting the entire thing including !expr, all with the same error above. 

Does it work for you? If so, then perhaps I should reinstall quarto or something else? And just to be clear, the `!expr` works no problem when I use it in chunk options like so

```r
{r}
#| fig.cap: !expr 1+1

hist(rnorm(1000))
```

**Reply by cscheid** [reply]
I apologize. `!expr` only appears to work inside `r` code cells, even though the knitr engine is the one running.

I don't know a clean way to do this in quarto.

**Reply by nurfatimaj** [reply]
I see. May I press a little more on this topic? So, the `!expr` doesn't work in the frontmatter, but this works
```yml
title: "`r 1+1`"
```
And to be more precise using the folder structure example above, it works when I render `website/research/paperA/index.qmd` file. Would you happen to know how can I make it work for rendering the listing page `website/listing.qmd` as well?

**Reply by mcanouil** [reply]
There was a Discussion about these constraints: <https://github.com/quarto-dev/quarto-cli/discussions/606#discussioncomment-2553877>

**Comment by mcanouil** [comment, marked as answer]
Here is an alternative, which basically includes the yaml (everything or only few selected fields) from another document within the Quarto document.

````md
---
format: html
params:
  project-path: "."
---

```{r}
#| output: asis
#| echo: false
cat(
  "---",
  yaml::as.yaml(
    rmarkdown::yaml_front_matter(
      input = list.files(
        path = params[["project-path"]],
        pattern = 'index.Rmd',
        full.names = TRUE
      )[c("title", "any_other_field")]
    )
  ),
  "---",
  sep = "\n"
)
```
````

**Reply by nurfatimaj** [reply]
Thank you! Unfortunately, it does not solve the issue with rendering the listing page, so for now I just have to remember to copy the yaml fields correctly. But I used your logic to copy other files that I want to include and link on the page. At least some automation 😊

**Reply by mcanouil** [reply]
Did you try to use a`metadata.yml`?

https://github.com/quarto-dev/quarto-cli/discussions/1840#discussioncomment-3383992

https://quarto.org/docs/websites/website-listings-custom.html
